### PR TITLE
sql/logictest: Added subtests to logic tests.

### PR DESCRIPTION
**Pull request by BramGruneir**
While working on the cascading deletes, I often found that I wanted to run subsections of the test file and the only way of doing so was to cut the test file to shreds. Normally, the logic tests run quickly, but when you're adding a large amount of logging, that's not the case and it becomes difficult to isolate the specific commands you want to run. But adding subtests, independent subsections of the logic tests can be run in isolation.

- Added a new command `subtest` to logic tests.
- Added a new variable SUBTESTS to our make file, which is only for use with testlogic.

Some notes:
- A subtest is defined by any section after a subtest command until it encounters another subtest comamnd or the end of file.
- All testing before encountering a subtest command is now a subtest called "base".

Release note: testing - SQL logic lests now support subtests.

### CODE REVIEW

**Comment by RaduBerinde** [comment]
:lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/logic_test.go, line 965 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05nCg59sDcufPNVkKl:-L05nCg59sDcufPNVkKm:b-2nb6ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/pkg/sql/logictest/logic_test.go#L965)):*
> ```Go
> 	s := newLineScanner(file, t.varMap)
> 	for s.Scan() {
> 		first := true
> ```

[nit] add a comment explaining this. Also, maybe `skipLine` would be a better name

---

*[pkg/sql/logictest/logic_test.go, line 972 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05mtW081TFQE7R1svM:-L05mtW081TFQE7R1svN:b-cdmbhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/pkg/sql/logictest/logic_test.go#L972)):*
> ```Go
> 		if len(fields) > 0 && fields[0] == "subtest" {
> 			if len(fields) != 2 {
> 				return errors.Errorf("%s:%d expected on one field following the subtest command", path, s.line)
> ```

one field

---

*[pkg/sql/logictest/logic_test.go, line 1006 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05nj5hrLU-ef-uYzeW:-L05nj5hrLU-ef-uYzeX:b-cdmbhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/pkg/sql/logictest/logic_test.go#L1006)):*
> ```Go
> 					if len(fields) != 2 {
> 						subtestError = errors.Errorf(
> 							"%s:%d expected on one field following the subtest command",
> ```

one field

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-:-L05nuJnr487WV3S4SPL:bnfp4nl)*

File: Makefile [hunk 1]
```diff
@@ -47,6 +47,10 @@ BENCHES :=
 ## Space delimited list of logic test files to run, for make testlogic.
 FILES :=
 
+## Regex for matching logic test subtests. This is always matched after files if
```
Comment (benesch) [inline Makefile]: nit if you're going through CI again anyway: "FILES".

**Comment by BramGruneir** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[Makefile, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05hzbc4jVavT81kn8_-r1-50:-L068A4_zyFXymVJCg-z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/Makefile#L50)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

nit if you're going through CI again anyway: "FILES". 
</blockquote></details>

Done.

---

*[pkg/sql/logictest/logic_test.go, line 965 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05nCg59sDcufPNVkKl:-L069PYHKwTlgWI-xFp2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/pkg/sql/logictest/logic_test.go#L965)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment explaining this. Also, maybe `skipLine` would be a better name
</blockquote></details>

Done.

---

*[pkg/sql/logictest/logic_test.go, line 972 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05mtW081TFQE7R1svM:-L069Z8ZcjMBpT1BMFXp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/pkg/sql/logictest/logic_test.go#L972)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

one field
</blockquote></details>

Done.

---

*[pkg/sql/logictest/logic_test.go, line 1006 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615#-L05nj5hrLU-ef-uYzeW:-L069ZYH865MLU5ZrNdJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13bf80d7c888ddc669e6dd5ec669a629ca1e13bd/pkg/sql/logictest/logic_test.go#L1006)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

one field
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20615)*

**Comment by RaduBerinde** [comment]
It seems that the many occurrences of
```go
  subtestError = errors.Errorf(...)
  subtestT.Error(subtestError)
  return
```
could be just
```go
  subtestT.Fatalf(...)
```

Either way, when one of these happens, we will keep running the statements that follow in a new subtest with the same name (?!)  
```
            --- PASS: TestLogic/5node-distsql/distsql_stats/base (0.00s)
            --- FAIL: TestLogic/5node-distsql/distsql_stats/lol (0.00s)
            	logic_test.go:1042: testdata/logic_test/distsql_stats:5 invalid repeat line: strconv.Atoi: parsing "lol": invalid syntax
            --- FAIL: TestLogic/5node-distsql/distsql_stats/lol#01 (0.00s)
            	logic_test.go:1042: testdata/logic_test/distsql_stats:6 invalid repeat line: strconv.Atoi: parsing "lol": invalid syntax
```

We probably want to stop the entire file (call Fatalf on `t.t` not `subtestT`?).

**Comment by BramGruneir** [comment]
You're right, there's a missing return that I'm sure I had in there.  I'll
submit a follow up PR today.

**Comment by RaduBerinde** [comment]
By the way, the new code is very hard to understand when you get into the details (sorry for not thinking about the details carefully when reviewing). There are many subtleties as to how subtests keep working even when previous subtests aren't run (in fact I was convinced it's broken a bunch of times, and had to "reverse debug" some tests to see how it works).

In particular, we are abusing `Run()` - if we have a `subtest a` that isn't run, we actually keep calling `Run("a")` for each statement until the next subtest (generating a ton of subtests `a#01, a#02` etc). This is pretty sketchy, and it can leak out, e.g.

```
$ make test PKG=./pkg/sql/logictest TESTS='Logic/default/cascade/0' TESTFLAGS=-v
        --- PASS: TestLogic/default/cascade (2.88s)
            --- PASS: TestLogic/default/cascade/base#01 (0.00s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_Basic#01 (0.29s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_PrimaryKeys#01 (0.31s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_CompositeFKs#01 (0.30s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_Restrict#01 (0.38s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_Interleaved#01 (0.43s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_InterleavedRestrict#01 (0.40s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_SelfReference#01 (0.03s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_SelfReferenceCycle#01 (0.03s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_TwoTableLoop#01 (0.12s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_TwoTableLoopCycle#01 (0.12s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_DoubleSelfReference#01 (0.39s)
```

We can get even get a bogus error:
```
    --- FAIL: TestLogic/default (0.01s)
        --- FAIL: TestLogic/default/cascade (2.20s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_Basic#05 (0.27s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_PrimaryKeys#05 (0.31s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_CompositeFKs#05 (0.30s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_Restrict#05 (0.37s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_Interleaved#05 (0.45s)
            --- PASS: TestLogic/default/cascade/DeleteCascade_InterleavedRestrict#05 (0.42s)
            --- FAIL: TestLogic/default/cascade/DeleteCascade_SelfReference#05 (0.00s)
            	logic_test.go:1406: testdata/logic_test/cascade:564: unknown command: id
        	logic_test.go:2063: 
        		testdata/logic_test/cascade:574: 
        		expected success, but found
        		pq: relation "self" does not exist
            --- FAIL: TestLogic/default/cascade/DeleteCascade_SelfReference#15 (0.00s)
            	logic_test.go:1088: 
            		testdata/logic_test/cascade:574: error in statement, skipping to next file
            --- FAIL: TestLogic/default/cascade/DeleteCascade_SelfReference#25 (0.00s)
            	logic_test.go:1000: testdata/logic_test/cascade:586: too many errors encountered, skipping the rest of the input
```

It might be cleaner to just read in the whole file into a `[]string` and then break that up into subtests beforehand.

We could also require that if a file uses `subtest`, it should start with that. Then we can avoid having the "base" subtest everywhere.

**Comment by RaduBerinde** [comment]
@BramGruneir are you still working on fixing this? If not, I can look at it later this week.

**Comment by BramGruneir** [comment]
I've started work on it on Friday and will have a PR ready this week.

**Pull request merged**